### PR TITLE
proof generation: batch normalize commitments for efficient transcript appending

### PR DESCRIPTION
**Pull request by jsign**
This PR creates a new helper method, `BatchNormalize([]*Element)` which is now used when generating a multiproof.

The reason is that the list `Cs` are projective points, thus when iterating over them and adding them to the transcript, the underlying `.Bytes()` will have to normalize them. That means a linear number of field inversions.

What we do now leverage the new `BatchNormalize(...)` method to normalize the projective points so when they're added to the transcript, the underlying `.Bytes()` (converting to affine) will go into fast paths avoiding inversions.

Note this situation happens for proof generation since the client has a point already in memory. For verifiers, this shouldn't be a problem since they would probably have the points in projective but deserialized from the proof, which is already normalized.

### CODE REVIEW

**Comment by jsign** [comment]
I found the problem, and it's actually in this PR!

The situation is that we might receive duplicated commitments, so we're "double normalizing" which is incorrect.
I'll fix soon.

File: banderwagon/element.go [hunk 1]
```diff
@@ -42,6 +42,41 @@ func (p Element) Bytes() [sizePointCompressed]byte {
...
+			if dedupedElements[i] == e {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+		dedupedElements = append(dedupedElements, e)
+	}
```
Comment (jsign) [inline banderwagon/element.go:61]: Deduping by doing lookups in a slice compared to a `map[*Element]struct{}`. I think we shouldn't have so many elements to justify a `map`.

**Comment by jsign** [comment]
Now things are good, ready for review again.

File: banderwagon/element.go (hunk 1 above)
Comment (kevaundray) [reply banderwagon/element.go:61]: Seems the gnark-crypto method does not do this: https://github.com/ConsenSys/gnark-crypto/blob/441dc0ffe639294b8d09e394f24ba7575577229c/ecc/bls12-381/g1.go#L920

So we may need to keep this even if we switch to upstream gnark-crypto

File: banderwagon/element.go [hunk 2]
```diff
@@ -42,6 +42,41 @@ func (p Element) Bytes() [sizePointCompressed]byte {
 	return x.Bytes()
 }
 
+func BatchNormalize(elements []*Element) {
```
Comment (kevaundray) [inline banderwagon/element.go:45]: Just for good measure, could you add a test for the point at infinity and duplicated elements

Comment (jsign) [reply banderwagon/element.go:45]: Done!

File: banderwagon/element_test.go [hunk 3]
```diff
@@ -224,3 +224,90 @@ func TestMultiMapToBaseField(t *testing.T) {
 		panic("expected scalar for point `A` is incorrect ")
 	}
 }
+
+func TestBatchNormalize(t *testing.T) {
+	t.Parallel()
+
+	t.Run("three points", func(t *testing.T) {
```
Comment (jsign) [inline banderwagon/element_test.go:231]: Basic test.

File: banderwagon/element_test.go [hunk 4]
```diff
@@ -224,3 +224,90 @@ func TestMultiMapToBaseField(t *testing.T) {
...
+
+		var A, B Element
+		A.Add(&Generator, &Generator)
+		B.Double(&A)
+
+		var expectedA, expectedB Element
+		expectedA.Set(&A).Normalise()
+		expectedB.Set(&B).Normalise()
+
+		BatchNormalize([]*Element{&A, &A, &B, &A})
```
Comment (jsign) [inline banderwagon/element_test.go:273]: Test duplicated pointers not being "over normalized" (i.e: dividing by `Z` more times than needed).

File: banderwagon/element_test.go [hunk 5]
```diff
@@ -224,3 +224,90 @@ func TestMultiMapToBaseField(t *testing.T) {
...
+
+		if !A.Equal(&expectedA) {
+			t.Fatal("expected point `A` is incorrect ")
+		}
+
+		if !B.inner.X.Equal(&expectedB.inner.X) ||
+			!B.inner.Y.Equal(&expectedB.inner.Y) ||
+			!B.inner.Z.Equal(&expectedB.inner.Z) {
+			t.Fatal("expected point `B` is incorrect ")
+		}
```
Comment (jsign) [inline banderwagon/element_test.go:311]: This feels a bit odd, but `.Equals(..)` always returns false if `Z == 0`. 
So here I'm being more explicit about `BatchNormalize` leading to the same projective coordinates as `Normalise` (one by one). 

Does this look fine?

Comment (kevaundray) [reply banderwagon/element_test.go:311]: Yep I think this is fine

**Review by kevaundray — APPROVED** [review]

**Comment by jsign** [comment]
@kevaundray, did a quick rebase to latest `master`.

**Review by kevaundray — APPROVED** [review]
